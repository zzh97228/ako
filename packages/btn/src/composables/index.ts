import { makeToggleGroup } from '@lagabu/shared';
const { innerSymbol: BtnToggleSymbol, useGroupProvider, useGroupConsumer, namespace } = makeToggleGroup(
  'BtnToggleGroup'
);

export { BtnToggleSymbol, useGroupConsumer, useGroupProvider, namespace };
