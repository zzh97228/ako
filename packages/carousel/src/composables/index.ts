import {
  clamp,
  convertToNumber,
  convertToUnit,
  deepEqual,
  isBool,
  isNumber,
  isString,
  useModel,
  setActive,
} from '@lagabu/shared';
import vue, {
  ExtractPropTypes,
  inject,
  InjectionKey,
  onBeforeUnmount,
  onMounted,
  provide,
  Ref,
  ref,
  SetupContext,
  toRef,
} from 'vue';
/**
 * @public
 */
export const CarouselSymbol: InjectionKey<{
  register: (...args: any[]) => any;
  unregister: (...args: any[]) => any;
  setParentHeight: (val: number | string | null) => any;
  disabled: Ref<boolean>;
  transitionName: Ref<string>;
  transitionCount: Ref<number>;
  inCarousel: boolean;
}> = Symbol('Carousel');
const defaultTimeout = 3000;
/**
 * @public
 */
export const carouselProps = {
  modelValue: [String, Number],
  disabled: Boolean,
  loop: Boolean,
  auto: [Boolean, String, Number],
};

function normalizeValue(str: any): number {
  let num: string | number = str;
  if (isNumber(num)) {
    return num;
  } else if (isString(str) && !isNaN((num = convertToNumber(str)))) {
    return num;
  }
  return 0;
}

function isValidIndex(idx: number, itemsLength: number) {
  return idx >= 0 && idx < itemsLength;
}
/**
 * @public carousel provider
 * @param props - carousel props
 */
export function useCarouselProvider(props: ExtractPropTypes<typeof carouselProps>) {
  const notAllowed = toRef(props, 'disabled');
  const transitionNextName = 'carousel-transition',
    transitionPrevName = 'carousel-transition--reverse',
    transitionName = ref(transitionNextName);
  const transitionHeight: Ref<string | null> = ref(null),
    transitionCount = ref(0);
  let timerId: any = 0;

  const items: Array<{
    isActive: Ref<boolean>;
    uid: number;
  }> = [];

  function setLoopValue(val: number) {
    if (props.loop) {
      val = val % items.length;
      val < 0 && (val += items.length);
    }
    return val;
  }

  const { model, lazyState, setInnerState } = useModel(props, notAllowed, (innerState, newVal, oldVal) => {
    newVal = normalizeValue(newVal);
    newVal = clamp(setLoopValue(newVal), 0, clamp(items.length - 1, 0));
    if (deepEqual(newVal, oldVal) || deepEqual(newVal, innerState.value) || props.disabled) return;
    innerState.value = newVal;
    transitionName.value = newVal > Number(oldVal) ? transitionNextName : transitionPrevName;
    updateItems(newVal);
  });

  function updateItems(activeIdx: number) {
    if (props.disabled) return;
    for (let i = 0; i < items.length; i++) {
      setActive(items[i].isActive, i === activeIdx);
    }
  }

  function register(isActive: Ref<boolean>, uid: number) {
    const idx =
      items.push({
        isActive,
        uid,
      }) - 1;
    setActive(items[idx].isActive, idx === lazyState.value);
  }

  function unregister(uid: number) {
    const idx = items.findIndex((item) => item.uid === uid);
    if (idx >= 0) {
      items.splice(idx, 1);
    }
  }

  function click(next?: boolean) {
    if (props.disabled) return;
    transitionName.value = next ? transitionNextName : transitionPrevName;
    let current = normalizeValue(lazyState.value);
    current = current + (next ? 1 : -1);
    current = clamp(setLoopValue(current), 0, clamp(items.length - 1, 0));
    updateItems(current);
    model.value = current;
  }

  function clickNext(e: Event) {
    if (props.disabled) {
      e.preventDefault();
      e.stopImmediatePropagation();
      return;
    }
    click(true);
    // reset auto timer;
    setAutoTimer();
  }

  function clickPrev(e: Event) {
    if (props.disabled) {
      e.preventDefault();
      e.stopImmediatePropagation();
      return;
    }
    click();
    // reset auto timer;
    setAutoTimer();
  }

  function setParentHeight(val: number | string | null) {
    const converted = convertToUnit(val) || null;
    if (converted === transitionHeight.value) return;
    transitionHeight.value = converted;
  }

  // initial lazyState and  active children
  setInnerState(normalizeValue(lazyState.value));
  // reset items when value invalid
  function initialValue() {
    let idx = lazyState.value as number;
    if (isValidIndex(idx, items.length)) return;
    idx = clamp(setLoopValue(idx), 0, clamp(items.length - 1, 0));
    setInnerState(idx);
    updateItems(idx);
  }
  function clearTimerId() {
    if (!timerId) return;
    clearInterval(timerId);
    timerId = 0;
  }

  function setAutoTimer() {
    clearTimerId();
    if (!props.auto) return;
    let timeout = defaultTimeout,
      tempTimeout: number;
    if (!isBool(props.auto) && (tempTimeout = convertToNumber(props.auto))) {
      timeout = tempTimeout;
    }
    timerId = setInterval(() => {
      if (lazyState.value === items.length - 1) {
        return clearTimerId();
      }
      click(true);
    }, timeout);
  }

  onMounted(() => {
    initialValue();
    setAutoTimer();
  });

  onBeforeUnmount(() => {
    clearTimerId();
  });

  provide(CarouselSymbol, {
    register,
    unregister,
    transitionName,
    transitionCount,
    disabled: notAllowed,
    inCarousel: true,
    setParentHeight,
  });

  return {
    clickNext,
    clickPrev,
    click,
    transitionHeight,
  };
}
let _cuid = 0;
/**
 * @public carousel consumer
 * @param isActive - Ref<boolean>
 */
export function useCarouselConsumer(isActive: Ref<boolean | undefined>) {
  const _id = _cuid++;
  const provider = inject(CarouselSymbol, {
    register: function (...args: any[]) {},
    unregister: function (...args: any[]) {},
    disabled: ref(false),
    transitionCount: ref(0),
    transitionName: ref(''),
    inCarousel: false,
    setParentHeight: function (val: number | string | null) {},
  });

  provider.inCarousel && provider.register(isActive, _id);

  onBeforeUnmount(() => {
    provider.inCarousel && provider.unregister(_id);
  });

  return {
    disabled: provider.disabled,
    transitionName: provider.transitionName,
    setParentHeight: provider.setParentHeight,
    transitionCount: provider.transitionCount,
  };
}
