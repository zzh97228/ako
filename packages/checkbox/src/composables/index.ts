import { makeToggleGroup, genGroupProps } from '@lagabu/shared';

const { innerSymbol: CheckboxSymbol, useGroupConsumer, useGroupProvider, namespace } = makeToggleGroup('CheckboxGroup');
export { CheckboxSymbol, useGroupConsumer, useGroupProvider, namespace, genGroupProps };
