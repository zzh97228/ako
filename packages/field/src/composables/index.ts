import { inject, InjectionKey, onBeforeUnmount, provide } from 'vue';

export const FieldGroupSymbol: InjectionKey<{
  register: (...args: any) => any;
  unregister: (...args: any) => any;
  inFieldGroup: boolean;
}> = Symbol('FieldGroup');
export function useFieldGroupProvider() {
  const fieldChild: any[] = [];

  const register = () => {};
  const unregister = () => {};

  function validate() {}

  provide(FieldGroupSymbol, {
    register,
    unregister,
    inFieldGroup: true,
  });
}

export function useFieldGroupConsumer() {
  const parent = inject(FieldGroupSymbol, {
    register: function () {},
    unregister: function () {},
    inFieldGroup: false,
  });

  if (parent.inFieldGroup) {
    parent.register();

    onBeforeUnmount(() => {
      parent.unregister();
    });
  }
}
