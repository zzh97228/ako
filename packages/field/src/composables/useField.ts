import { inject, InjectionKey, onBeforeUnmount, provide, watch, WatchStopHandle } from 'vue';
import { deepEqual, isUndefined, ModelReturn } from '@lagabu/shared';
export const FieldSymbol: InjectionKey<{
  inField: boolean;
  register: (instance: FieldContent) => any;
  unregister: (instance: FieldContent) => any;
}> = Symbol('Field');

let fieldUID = 0;
export const enum FieldEnum {
  parent = 1,
  child = 1 << 1,
}
export type FieldContentOptions<T> = { [props: string]: any } & ModelReturn<T>;
export class FieldContent<T = any> {
  fieldUID: number;
  lazyState: ModelReturn<T>['lazyState'];
  model: ModelReturn<T>['model'];
  type: FieldEnum;

  childStop?: WatchStopHandle;
  currentStop?: WatchStopHandle;
  constructor(opts: FieldContentOptions<T>, type: FieldEnum = FieldEnum.child) {
    this.type = type;
    this.fieldUID = fieldUID++;
    this.model = opts.model;
    this.lazyState = opts.lazyState;
  }

  bindChild(child: FieldContent<T>) {
    if (this.type !== FieldEnum.parent && child.type !== FieldEnum.child) return;
    // set init value
    if (!isUndefined(this.lazyState.value)) {
      child.lazyState.value = this.lazyState.value;
    } else if (!isUndefined(child.lazyState.value)) {
      this.lazyState.value = child.lazyState.value;
    }

    // watch child's model and emit model
    this.childStop = watch(child.model, (newVal, oldVal) => {
      if (deepEqual(newVal, oldVal) || deepEqual(this.lazyState.value, newVal)) return;
      this.model.value = newVal;
    });
    // watch current lazyState input and change child's input
    this.currentStop = watch(
      () => this.lazyState.value,
      (newVal, oldVal) => {
        if (deepEqual(newVal, oldVal) || deepEqual(child.lazyState.value, newVal)) return;
        child.lazyState.value = newVal;
      }
    );
  }

  unregister() {
    this.childStop && this.childStop();
    this.currentStop && this.currentStop();
  }
}

export function useFieldProvider<T extends unknown>(modelOptions: ModelReturn<T>) {
  const parent = new FieldContent(modelOptions, FieldEnum.parent);
  function register(child: FieldContent) {
    parent.bindChild(child);
  }
  function unregister(child: FieldContent) {
    parent.unregister();
    child.unregister();
  }
  provide(FieldSymbol, {
    register,
    unregister,
    inField: true,
  });

  return {};
}
export function useFieldInjector<T extends unknown>(modelOptions: ModelReturn<T>) {
  const content = new FieldContent(modelOptions);
  const field = inject(FieldSymbol, {
    inField: false,
    register: (...args: any[]) => {},
    unregister: (...args: any[]) => {},
  });
  if (field.inField) {
    field.register(content);

    onBeforeUnmount(() => {
      field.unregister(content);
    });
  }
}
