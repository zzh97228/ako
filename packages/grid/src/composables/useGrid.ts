import { convertToUnit } from '@lagabu/shared';
import { FlexEnum } from '../services';
import vue, { computed, ExtractPropTypes, Ref, inject, InjectionKey, provide, toRef, ref } from 'vue';

export const FLEX_KEYS: Array<string | FlexEnum> = [FlexEnum.xs, FlexEnum.sm, FlexEnum.md, FlexEnum.lg, FlexEnum.xl];
export const GridSymbol: InjectionKey<{
  gutter: Ref<string | null>;
  column: Ref<boolean>;
}> = Symbol('Grid');
export const gridProps = {
  gutter: {
    type: String,
    default: null,
  },
  column: Boolean,
};

export function useGridProvider(props: ExtractPropTypes<typeof gridProps>) {
  const classes = computed(() => {
    if (!FLEX_KEYS.includes(props.gutter)) return {};
    return {
      [`row--gutter-${props.gutter}${props.column ? '-column' : ''}`]: !!props.gutter,
    };
  });
  const styles = computed(() => {
    const style: Record<string, string> = {};
    if (!props.gutter || FLEX_KEYS.includes(props.gutter)) return style;
    let value = convertToUnit(props.gutter) || '0px';
    if (!value.match(/^\-/)) {
      value = '-' + value;
    }
    if (!props.column) {
      style['margin-left'] = value;
      style['margin-right'] = value;
    } else {
      style['margin-top'] = value;
      style['margin-bottom'] = value;
    }
    return style;
  });
  provide(GridSymbol, {
    gutter: toRef(props, 'gutter'),
    column: toRef(props, 'column'),
  });

  return {
    class: classes,
    style: styles,
  };
}

export function useGridConsumer() {
  const grid = inject(GridSymbol, {
    gutter: ref(null),
    column: ref(false),
  });
  const styles = computed(() => {
    let gutter: string | null = grid.gutter.value;
    if (!gutter || FLEX_KEYS.includes(gutter)) return {};
    gutter = convertToUnit(gutter) || '0px';
    return {
      [`padding-${grid.column.value ? 'top' : 'left'}`]: gutter,
      [`padding-${grid.column.value ? 'bottom' : 'right'}`]: gutter,
    };
  });

  return {
    style: styles,
  };
}
