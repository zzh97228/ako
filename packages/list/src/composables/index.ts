export const ListSymbol = Symbol('List');
export function useListProvider() {}

export function useListConsumer() {}
