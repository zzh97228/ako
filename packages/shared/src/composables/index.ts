export * from './useSize';
export * from './useColor';
export * from './useElevation';