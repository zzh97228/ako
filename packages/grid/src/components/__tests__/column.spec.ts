import { mount, VueWrapper } from '@vue/test-utils';
import { computed } from 'vue';
import { GridSymbol } from '../../composables/useGrid';
import Column from '../column';

describe('column.ts', () => {
  let mountFunc: (options?: object, column?: boolean) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (options = {}, column?: boolean) => {
      return mount(Column, {
        global: {
          provide: {
            [GridSymbol as symbol]: {
              gutter: computed(() => '4px'),
              column: computed(() => column),
            },
          },
        },
        ...options,
      });
    };
  });
  it('should render column component with default props', async () => {
    const wrapper = mountFunc({
      props: {
        cols: 6,
        grow: true,
      },
    });

    expect(wrapper.classes()).toContain('col');
    expect(wrapper.classes()).toContain('col-6');
    expect(wrapper.classes()).toContain('col--grow');

    await wrapper.setProps({
      xs: 6,
      md: 4,
    });
    expect(wrapper.classes()).toContain('col-6-xs');
    expect(wrapper.classes()).toContain('col-4-md');
    expect(wrapper.attributes('style')).toContain('padding-left: 4px');
    expect(wrapper.attributes('style')).toContain('padding-right: 4px');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should have vertical padding when row provide column props', () => {
    const wrapper = mountFunc(
      {
        props: {
          cols: 10,
          shrink: true,
        },
      },
      true
    );
    expect(wrapper.classes()).toContain('col--shrink');
    expect(wrapper.attributes('style')).toContain('padding-top: 4px');
    expect(wrapper.attributes('style')).toContain('padding-bottom: 4px');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
