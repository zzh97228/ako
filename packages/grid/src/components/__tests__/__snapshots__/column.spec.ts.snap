// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`column.ts should have vertical padding when row provide column props 1`] = `
<div class="col col--shrink col-10"
     style="padding-top: 4px; padding-bottom: 4px;"
>
</div>
`;

exports[`column.ts should render column component with default props 1`] = `
<div class="col col--grow col-6 col-6-xs col-4-md"
     style="padding-left: 4px; padding-right: 4px;"
>
</div>
`;
